### PR TITLE
Disable checkpointing, earlystopping and logging with fast_dev_run

### LINKED ISSUE #5136: fast_dev_run should set max_steps
## 🐛 Bug

`Trainer(fast_dev_run=True)` should set `max_steps` accordingly, but `max_steps` is `None`

## Please reproduce using [the BoringModel and post here](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing)

https://colab.research.google.com/drive/1PFoozULca2zFtw0Ljor5V8AAGsGqe1Y4#scrollTo=quj4LUDgmFvj
### Expected behavior

`max_steps` should be set to `1 if fast_dev_run is True else fast_dev_run`

### Environment

**Note**: `Bugs with code` are solved faster ! `Colab Notebook` should be made `public` !

* `IDE`: Please, use our python [bug_report_model.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py
) template.

* `Colab Notebook`: Please copy and paste the output from our [environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py) (or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```
* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.18.5
	- pyTorch_debug:     True
	- pyTorch_version:   1.7.0+cu101
	- pytorch-lightning: 1.1.0
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.6.9
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #5136 + disable checkpointing and logging when using `fast_dev_run`.

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes [if needed]?
- [x] Did you write any new necessary tests [no need for typos, docs]? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone are aligned!**
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/properties.py [hunk 1]
```diff
@@ -218,6 +218,20 @@ def weights_save_path(self) -> str:
...
+    def early_stopping_callback(self) -> Optional[ModelCheckpoint]:
+        """
+        The first early_stopping callback in the Trainer.callbacks list, or ``None`` if
+        no early_stopping callbacks exist.
+        """
+        callbacks = self.early_stopping_callbacks
+        return callbacks[0] if len(callbacks) > 0 else None
+
+    @property
+    def early_stopping_callbacks(self) -> List[ModelCheckpoint]:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/properties.py:231]:
```suggestion
    def early_stopping_callbacks(self) -> List[EarlyStopping]:
```

File: pytorch_lightning/trainer/connectors/debugging_connector.py [hunk 2]
```diff
@@ -54,11 +56,18 @@ def on_init_start(
             limit_train_batches = fast_dev_run
             limit_val_batches = fast_dev_run
             limit_test_batches = fast_dev_run
+            self.trainer.max_steps = fast_dev_run
             self.trainer.num_sanity_val_steps = 0
             self.trainer.max_epochs = 1
+            self.trainer.val_check_interval = 1.0
+            self.trainer.check_val_every_n_epoch = 1
+            self.trainer.logger = None
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/debugging_connector.py:64]: We may have to replace it with the dummy logger we use for multi-gpu training. If the user makes calls to `trainer.logger.experiment.custom_logging...` their code would brake when setting `fast_dev_run=True`. What do you think?

File: tests/trainer/flags/test_fast_dev_run.py [hunk 3]
```diff
@@ -19,3 +23,70 @@ def test_skip_on_fast_dev_run_tuner(tmpdir, tuner_alg):
...
+            self.training_step_called = True
+            return super().training_step(batch, batch_idx)
+
+        def validation_step(self, batch, batch_idx):
+            self.validation_step_called = True
+            return super().validation_step(batch, batch_idx)
+
+    def _make_fast_dev_run_assertions(trainer):
+        # there should be no logger with fast_dev_run
+        assert trainer.logger is None
```
Comment (awaelchli) [inline tests/trainer/flags/test_fast_dev_run.py:51]: see my comment about loggers above

File: tests/trainer/flags/test_fast_dev_run.py [hunk 4]
```diff
@@ -19,3 +23,70 @@ def test_skip_on_fast_dev_run_tuner(tmpdir, tuner_alg):
     expected_message = f'Skipping {tuner_alg} since fast_dev_run is enabled.'
     with pytest.warns(UserWarning, match=expected_message):
         trainer.tune(model)
+
+
+@pytest.mark.parametrize('fast_dev_run', [1, 4])
+@mock.patch.dict(os.environ, {"PL_DEV_DEBUG": "1"})
+def test_mc_and_logger_not_called_with_fastdevrun(tmpdir, fast_dev_run):
```
Comment (awaelchli) [inline tests/trainer/flags/test_fast_dev_run.py:30]:
```suggestion
def test_callbacks_and_logger_not_called_with_fastdevrun(tmpdir, fast_dev_run):
```

File: pytorch_lightning/trainer/connectors/debugging_connector.py [hunk 5]
```diff
@@ -54,11 +56,18 @@ def on_init_start(
...
+            self.trainer.callbacks = [
+                c for c in self.trainer.callbacks if not isinstance(c, (EarlyStopping, ModelCheckpoint))
+            ]
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/debugging_connector.py:67]: same here, rather than removing the callbacks, we may have to keep them and just disable them because otherwise it could break user code that accesses these. What do you think?

**Review by awaelchli — COMMENTED** [review]
nice! we just need to figure out a few details regarding callbacks and logger :)

File: pytorch_lightning/callbacks/progress.py [hunk 6]
```diff
@@ -323,7 +322,7 @@ def on_epoch_start(self, trainer, pl_module):
         super().on_epoch_start(trainer, pl_module)
         total_train_batches = self.total_train_batches
         total_val_batches = self.total_val_batches
-        if total_train_batches != float('inf') and not trainer.fast_dev_run:
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply pytorch_lightning/callbacks/progress.py:326]: yes.

File: pytorch_lightning/trainer/connectors/debugging_connector.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/debugging_connector.py]: valid point. Updated.

File: pytorch_lightning/trainer/connectors/debugging_connector.py (hunk 5 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/debugging_connector.py]: valid point. Updated.

File: pytorch_lightning/trainer/properties.py [hunk 7]
```diff
@@ -218,18 +218,38 @@ def weights_save_path(self) -> str:
...
         return self._weights_save_path
 
+    @property
+    def early_stopping_callback(self) -> Optional[EarlyStopping]:
+        """
+        The first :class:`~pytorch_lightning.callbacks.early_stopping.EarlyStopping`
+        callback in the Trainer.callbacks list, or ``None`` if it doesn't exist.
+        """
+        callbacks = self.early_stopping_callbacks
+        return callbacks[0] if len(callbacks) > 0 else None
```
Comment (Borda) [inline pytorch_lightning/trainer/properties.py:228]: why do we need this get for first callback?

File: pytorch_lightning/trainer/properties.py [hunk 8]
```diff
@@ -218,18 +218,38 @@ def weights_save_path(self) -> str:
...
     @property
     def checkpoint_callback(self) -> Optional[ModelCheckpoint]:
         """
-        The first checkpoint callback in the Trainer.callbacks list, or ``None`` if
-        no checkpoint callbacks exist.
+        The first :class:`~pytorch_lightning.callbacks.model_checkpoint.ModelCheckpoint`
+        callback in the Trainer.callbacks list, or ``None`` if it doesn't exist.
         """
         callbacks = self.checkpoint_callbacks
         return callbacks[0] if len(callbacks) > 0 else None
```
Comment (Borda) [inline pytorch_lightning/trainer/properties.py:245]:
```suggestion
        return callbacks[0] if callbacks > 0 else None
```

File: pytorch_lightning/trainer/properties.py [hunk 9]
```diff
@@ -218,18 +218,38 @@ def weights_save_path(self) -> str:
...
+    @property
+    def early_stopping_callbacks(self) -> List[EarlyStopping]:
+        """
+        A list of all instances of :class:`~pytorch_lightning.callbacks.early_stopping.EarlyStopping`
+        found in the Trainer.callbacks list.
+        """
+        return [c for c in self.callbacks if isinstance(c, EarlyStopping)]
+
     @property
     def checkpoint_callback(self) -> Optional[ModelCheckpoint]:
```
Comment (Borda) [inline pytorch_lightning/trainer/properties.py:239]: is this needed somewhere? :]

File: tests/trainer/flags/test_fast_dev_run.py [hunk 10]
```diff
@@ -19,3 +24,75 @@ def test_skip_on_fast_dev_run_tuner(tmpdir, tuner_alg):
     expected_message = f'Skipping {tuner_alg} since fast_dev_run is enabled.'
     with pytest.warns(UserWarning, match=expected_message):
         trainer.tune(model)
+
+
+@pytest.mark.parametrize('fast_dev_run', [1, 4])
+@mock.patch.dict(os.environ, {"PL_DEV_DEBUG": "1"})
+def test_callbacks_and_logger_not_called_with_fastdevrun(tmpdir, fast_dev_run):
```
Comment (Borda) [inline tests/trainer/flags/test_fast_dev_run.py:31]: is it just move from other file or can you pls tell us what is the diff?

File: pytorch_lightning/trainer/properties.py (hunk 8 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/properties.py:245]: callbacks is a list, `callbacks > 0` will raise an error.

File: tests/trainer/flags/test_fast_dev_run.py (hunk 10 above)
Comment (rohitgr7) [reply tests/trainer/flags/test_fast_dev_run.py:31]: not exactly. Before this PR, `ModelCheckpoint` was enabled with `fast_dev_run`, but now it isn't. Also updated the test to assert everything that is required to be disabled (callbacks and loggers) so thought it's better to keep this one along with other `fast_dev_run` tests.

File: pytorch_lightning/trainer/properties.py (hunk 7 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/properties.py:228]: it's similar to what we have for checkpoint_callback.

File: pytorch_lightning/trainer/properties.py (hunk 9 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/properties.py:239]: yes, used at some places in the codebase.

File: tests/trainer/flags/test_fast_dev_run.py [hunk 11]
```diff
@@ -19,3 +24,75 @@ def test_skip_on_fast_dev_run_tuner(tmpdir, tuner_alg):
...
+    Test that ModelCheckpoint, EarlyStopping and Logger are turned off with fast_dev_run
+    """
+    class FastDevRunModel(BoringModel):
+        def __init__(self):
+            super().__init__()
+            self.training_step_called = False
+            self.validation_step_called = False
+            self.test_step_called = False
+
+        def training_step(self, batch, batch_idx):
```
Comment (awaelchli) [inline tests/trainer/flags/test_fast_dev_run.py:42]: Let's also add a simple log call here, with self.log and also with self.logger.experiment.log?

Comment (rohitgr7) [reply tests/trainer/flags/test_fast_dev_run.py:42]: good point :+1:

File: tests/trainer/flags/test_fast_dev_run.py [hunk 12]
```diff
@@ -19,3 +25,77 @@ def test_skip_on_fast_dev_run_tuner(tmpdir, tuner_alg):
...
+            self.log('some_metric', torch.tensor(7.))
+            self.logger.experiment.add_scaler('some_distribution', torch.randn(7) + batch_idx)
```
Comment (awaelchli) [inline tests/trainer/flags/test_fast_dev_run.py:45]:
```suggestion
            self.logger.experiment.dummy_log('some_distribution', torch.randn(7) + batch_idx)
```
did you misspell "add_scalar" on purpose? let's make it very obvious that we deal with a dummy logger here :))

**Review by awaelchli — APPROVED** [review]
👍

File: pytorch_lightning/callbacks/lr_monitor.py [hunk 13]
```diff
@@ -105,15 +105,15 @@ def on_train_batch_start(self, trainer, *args, **kwargs):
             interval = 'step' if self.logging_interval is None else 'any'
             latest_stat = self._extract_stats(trainer, interval)
 
-            if trainer.logger is not None and latest_stat:
+            if latest_stat:
```
Comment (tchaton) [inline pytorch_lightning/callbacks/lr_monitor.py:108]: Why did you remove trainer.logger check ? What happens if Trainer(logger=False)

**Review by tchaton — APPROVED** [review]
Great work ! Can you clarify why you removed trainer.logger ?

File: pytorch_lightning/callbacks/lr_monitor.py (hunk 13 above)
Comment (rohitgr7) [reply pytorch_lightning/callbacks/lr_monitor.py:108]: we have a hard stop if there is no logger, so it's not required here.
https://github.com/PyTorchLightning/pytorch-lightning/blob/15a400b95faf45124f28fd8ca4ef7520d47df88a/pytorch_lightning/callbacks/lr_monitor.py#L75-L84

File: CHANGELOG.md [hunk 14]
```diff
@@ -20,6 +20,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 
 ### Fixed
+- Disabled checkpointing, earlystopping and logger with `fast_dev_run` ([#5277](https://github.com/PyTorchLightning/pytorch-lightning/pull/5277))
```
Comment (Borda) [inline CHANGELOG.md:23]:
```suggestion

- Disabled checkpointing, earlystopping and logger with `fast_dev_run` ([#5277](https://github.com/PyTorchLightning/pytorch-lightning/pull/5277))
```

**Review by Borda — APPROVED** [review]
lgtm

**Comment by Borda** [comment]
@rohitgr7 mind check isort formating?

**Comment by Borda** [comment]
just thinking if the disable checkpointing on fast_dev_run is rather a API change so it shall be 1.2?
cc: @tchaton @awaelchli

File: docs/source/trainer.rst [hunk 15]
```diff
@@ -666,9 +666,9 @@ Under the hood the pseudocode looks like this when running *fast_dev_run* with a
 .. note::
 
     This argument is a bit different from ``limit_train/val/test_batches``. Setting this argument will
-    disable tuner, logger callbacks like ``LearningRateLogger`` and runs for only 1 epoch. This must be
-    used only for debugging purposes. ``limit_train/val/test_batches`` only limits the number of batches and won't
-    disable anything.
+    disable tuner, checkpoint callbacks, early stopping callbacks, loggers and logger callbacks like
+    ``LearningRateLogger`` and runs for only 1 epoch. This must be used only for debugging purposes.
```
Comment (carmocca) [inline docs/source/trainer.rst:670]:
```suggestion
    ``LearningRateLogger`` and runs for only one step during the first epoch. This must be used only for debugging purposes.
```

(line will be too long, do not directly commit from suggestion)

Comment (rohitgr7) [reply docs/source/trainer.rst:670]: it can run multiple steps during an epoch if you set `fast_dev_run = int( > 1)`. Also it runs for 1 epoch only.

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 16]
```diff
@@ -478,14 +479,14 @@ def __resolve_ckpt_dir(self, trainer, pl_module):
             version, name = trainer.accelerator_backend.broadcast((version, trainer.logger.name))
 
             ckpt_path = os.path.join(
-                save_dir, name, version, "checkpoints"
+                save_dir, str(name), version, "checkpoints"
```
Comment (carmocca) [inline pytorch_lightning/callbacks/model_checkpoint.py:482]: why is `str` required?

Comment (rohitgr7) [reply pytorch_lightning/callbacks/model_checkpoint.py:482]: in DummyLogger version is None, so this worked out :smile:

**Review by carmocca — APPROVED** [review]

**Review by s-rog — APPROVED** [review]
